### ISSUE TITLE
Quickinfo incorrectly computes zero-candidate type parameter with contextual inference candidate

### ISSUE THREAD

**Issue by devanshj** [microsoft/TypeScript#44879]
# Bug Report
### 🔎 Search Terms

intellisense, contextual inference, call site inference
Related: #44821

### 🕗 Version & Regression Information

Tested with v4.3.4

### ⏯ Playground Link

[Playground](https://www.typescriptlang.org/play?jsx=0#code/LYCg3gBAZg9jBcEAkIBEsaoJQQL5YG4IB6YiAVwDsBTADwAdqBjAF2oBMJqAnbmbgFAD2zADYBDbtQhMYlAM4sIweAIgQAPAGUutNpXbyIi7gEtKAcwB8IeYkgBtAPoRzELQF17yRADkYAJKUUDzaVnh4OAC84QBuMKbsQiJMElIycoo+mloANBAAKrr6hu42doXRVmoQkEiIOrhCLACejBD+QSHcGgXhURAOBR5DxdQGRuKULRAA-BAADBCINLE8HgQCQA)

### 💻 Code

```ts
m({ foo: $("foo") }); // unexpected error

declare const m:
  <S extends string>(s: { [_ in S]: { $: NoInfer<S> } }) => void

declare const $: <S, T extends S>(s: T) =>
  { $: S }

type NoInfer<T> = [T][T extends any ? 0 : never];
```

### 🙁 Actual behavior

Code does not compile

### 🙂 Expected behavior

Code should compile. Especially when the intellisense gets the type parameters right and also provides completions.

Expected type parameters and return type -

![](https://user-images.githubusercontent.com/30295578/124342263-4911e980-dbe0-11eb-8784-c723b2ff23c5.png)

Expected completion -

![](https://user-images.githubusercontent.com/30295578/124342299-8a09fe00-dbe0-11eb-8bfb-182e33c970ca.png)

But doesn't compile and gives error - 

![](https://user-images.githubusercontent.com/30295578/124342325-ac038080-dbe0-11eb-9cd9-0b04511796fd.png)

**Comment by RyanCavanaugh** [comment]
The compile-time inference is correct. I'm not sure what `$` is trying to accomplish, but it's not written in a way that can produce meaningful results -- it says "I take a `T` and return *some supertype of that* in `$`", which doesn't really make any sense.

**Comment by devanshj** [comment]
Yeah it was sort of a workaround, it should have simply been `$: <T>(s: T) => T` as you said in [here](https://github.com/microsoft/TypeScript/issues/44821#issuecomment-875206249).
